concordium-smart-contract-testing/src/impls.rs

```diff
@@ -1065,11 +1387,245 @@ impl Chain {
+                    .await
+                    .map_err(|_| SetupExternalNodeError::CheckQueryBlockTimeout)??;
+            Ok::<_, SetupExternalNodeError>((client, checked_query_block))
+        })?;
+
+        // Set or replace the node connection.
+        self.external_node_connection = Some(ExternalNodeConnection {
+            client,
+            runtime,
+            query_block: checked_query_block,
+        });
+
+        Ok(())
+    }
+
+    /// Try to get a mutable reference to [`ExternalNodeConnection`] or return
+    /// an error.
+    ///
+    /// The connection is only available, if the [`Chain`] has been set up with
+    /// an external node connection via
+    /// [`ChainBuilder::external_node_connection`] in the [`ChainBuilder`].
+    fn external_node_connection_mut(
+        &mut self,
+    ) -> Result<&mut ExternalNodeConnection, ExternalNodeNotConfigured> {
+        match &mut self.external_node_connection {
+            None => Err(ExternalNodeNotConfigured),
+            Some(data) => Ok(data),
+        }
+    }
+
+    /// Try to get an immutable reference to [`ExternalNodeConnection`] or
+    /// return an error.
+    ///
+    /// The connection is only available, if the [`Chain`] has been set up with
+    /// an external node connection via
+    /// [`ChainBuilder::external_node_connection`] in the [`ChainBuilder`].
+    fn external_node_connection(
+        &self,
+    ) -> Result<&ExternalNodeConnection, ExternalNodeNotConfigured> {
+        match &self.external_node_connection {
+            None => Err(ExternalNodeNotConfigured),
+            Some(data) => Ok(data),
+        }
+    }
+
     /// Return the current microCCD per euro exchange rate.
     pub fn micro_ccd_per_euro(&self) -> ExchangeRate { self.parameters.micro_ccd_per_euro }
 
     /// Return the current euro per energy exchange rate.
     pub fn euro_per_energy(&self) -> ExchangeRate { self.parameters.euro_per_energy }
+
+    /// Return the current block time.
+    pub fn block_time(&self) -> Timestamp { self.parameters.block_time }
+
+    /// Return the block used for external queries by default.
+    ///
+    /// The block can be set with [`ChainBuilder::external_query_block`] when
+    /// building the [`Chain`].
+    ///
+    /// This method returns an error if the external node has not been
+    /// configured.
+    pub fn external_query_block(&self) -> Result<BlockHash, ExternalNodeNotConfigured> {
+        self.external_node_connection().map(|conn| conn.query_block)
+    }
+}
+
+impl ExternalNodeConnection {
+    /// Execute an async task with the [`sdk::v2::Client`].
+    ///
+    /// If a block is provided, it will be used for the query. Otherwise, it
+    /// will use the default query block.
+    ///
+    /// If the task takes longer than [`EXTERNAL_NODE_TIMEOUT_DURATION`] then
+    /// the connection times out and an [`Err(ExternalNodeError::Timeout)`] is
+    /// returned.
+    fn with_client<T, F, Fut>(
```

Comment:
I think it might be useful to clarify here taht this function cannot be nested, that'll lead to a panic.